### PR TITLE
H12media Bid Adapter: added new optional params, fixes, and various integration support

### PR DESCRIPTION
**Pull request by h12media**
## Type of change

- [x] Bugfix
- [x] Feature
- [ ] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change

- Separate bid requests
- Added new optional params: pubsubid and pubcontainerid
- Fixed UserSync requests
- Added withCredentials support
- Added USP support
- Added UserId support
- Other small fixes

- contact email of the adapter’s maintainer: contact@h12-media.com
- [x] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:
- Added new optional params: pubsubid and pubcontainerid

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/
https://github.com/prebid/prebid.github.io/pull/2781

### CODE REVIEW

**Comment by ChrisHuie** [comment]
Docs pr that accompanies this -> https://github.com/prebid/prebid.github.io/pull/2781

File: modules/h12mediaBidAdapter.js [hunk 1]
```diff
@@ -206,4 +217,32 @@ function getLocalDateFormatted() {
   return `${d.getFullYear()}-${two(d.getMonth() + 1)}-${two(d.getDate())} ${two(d.getHours())}:${two(d.getMinutes())}:${two(d.getSeconds())}`;
 }
 
+function inIframe () {
```
Comment (aleksatr) [inline modules/h12mediaBidAdapter.js:220]: why didn't you utilize `utils.inIframe()` ?

File: modules/h12mediaBidAdapter.js [hunk 2]
```diff
@@ -16,21 +15,30 @@ export const spec = {
...
     const docSize = getDocumentDimensions();
 
-    const bidrequests = validBidRequests.map((bidRequest) => {
+    return validBidRequests.map((bidRequest) => {
       const bidderParams = bidRequest.params;
-      const adUnitElement = document.getElementById(bidRequest.adUnitCode);
+      const requestUrl = bidderParams.endpointdom || DEFAULT_URL;
+      let pubsubid = bidderParams.pubsubid || '';
+      if (pubsubid && pubsubid.length > 32) {
+        utils.logError('Bidder param \'pubsubid\' should be less than 32 chars.');
```
Comment (aleksatr) [inline modules/h12mediaBidAdapter.js:27]: judging by your condition, it can also be `equal` to 32 chars.

File: modules/h12mediaBidAdapter.js [hunk 3]
```diff
@@ -16,21 +15,30 @@ export const spec = {
...
+        pubsubid = '';
+      }
+      const pubcontainerid = bidderParams.pubcontainerid;
+      const adUnitElement = document.getElementById(pubcontainerid || bidRequest.adUnitCode);
       const ishidden = !isVisible(adUnitElement);
-      const coords = {
+      const coords = isiframe ? {
+        x: getFramePos()[0],
+        y: getFramePos()[1],
```
Comment (aleksatr) [inline modules/h12mediaBidAdapter.js:35]: why calling `getFramePos()` twice, why don't you save the returned value in some variable `pos` and use it later as `pos[0]` and `pos[1]`

File: modules/h12mediaBidAdapter.js [hunk 4]
```diff
@@ -40,63 +48,67 @@ export const spec = {
...
+        method: 'POST',
+        url: requestUrl,
+        options: {withCredentials: true},
+        data: {
+          gdpr: !!utils.deepAccess(bidderRequest, 'gdprConsent.gdprApplies', false),
+          gdpr_cs: utils.deepAccess(bidderRequest, 'gdprConsent.consentString', ''),
+          usp: !!utils.deepAccess(bidderRequest, 'uspConsent', false),
+          usp_cs: utils.deepAccess(bidderRequest, 'uspConsent', ''),
+          topLevelUrl: utils.deepAccess(bidderRequest, 'refererInfo.referer', ''),
+          refererUrl: window.top.document.referrer || window.document.referrer,
```
Comment (aleksatr) [inline modules/h12mediaBidAdapter.js:65]: you should guard `window.top` access with a `try-catch`, same goes for other places where you use it.

File: modules/h12mediaBidAdapter.js (hunk 2 above)
Comment (h12media) [reply modules/h12mediaBidAdapter.js]: Fixed, the log message is changed

File: modules/h12mediaBidAdapter.js (hunk 3 above)
Comment (h12media) [reply modules/h12mediaBidAdapter.js]: Fixed, added cache variable

File: modules/h12mediaBidAdapter.js (hunk 4 above)
Comment (h12media) [reply modules/h12mediaBidAdapter.js]: Fixed, `try-catch` with `window.top` added to all places where it was missing

File: modules/h12mediaBidAdapter.js (hunk 1 above)
Comment (h12media) [reply modules/h12mediaBidAdapter.js]: Fixed, now `utils.inIframe()` used

**Pull request merged**